repo/migrations/Migration012_test.go

```diff
@@ -0,0 +1,564 @@
+	fragments = append([]string{testMigration012_datadir}, fragments...)
+	return path.Join(fragments...)
+}
+
+func testMigration012_assertListingsMigratedCorrectly(t *testing.T) {
+	for _, listing := range testMigraion012_listingFixtures {
+		testMigration012_assertListingMigratedCorrectly(t, listing)
+	}
+}
+
+func testMigration012_assertListingMigratedCorrectly(t *testing.T, listingBeforeMigration pb.SignedListing) {
+	listingPath := testMigration012_filepath("root", "listings", listingBeforeMigration.Listing.Slug+".json")
+	listingJSON, err := ioutil.ReadFile(listingPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	sl := new(pb.SignedListing)
+	err = jsonpb.UnmarshalString(string(listingJSON), sl)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Check version
+	if sl.Listing.Metadata.Version != 4 {
+		t.Fatal("Incorrect version.\nWanted:", 4, "\nGot:", sl.Listing.Metadata.Version)
+	}
+
+	// Check signature
+	err = testMigration012_verifySignature(
+		sl.Listing,
+		sl.Listing.VendorID.Pubkeys.Identity,
+		sl.Signature,
+		sl.Listing.VendorID.PeerID,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// check other data is the same
+}
+
+func testMigration012_verifySignature(msg proto.Message, pk []byte, signature []byte, peerID string) error {
+	ser, err := proto.Marshal(msg)
+	if err != nil {
+		return err
+	}
+	pubkey, err := crypto.UnmarshalPublicKey(pk)
+	if err != nil {
+		return err
+	}
+	valid, err := pubkey.Verify(ser, signature)
+	if err != nil {
+		return err
+	}
+	if !valid {
+		return errors.New("Invalid signature")
+	}
+	pid, err := peer.IDB58Decode(peerID)
+	if err != nil {
+		return err
+	}
+	if !pid.MatchesPublicKey(pubkey) {
+		return errors.New("Pubkey does not match peer ID")
+	}
+	return nil
+}
+
+func testMigration012_assertListingIndexMigratedCorrectly(t *testing.T) {
+	listingsIndexJSON, err := ioutil.ReadFile(testMigraion012_listingsIndexPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if string(listingsIndexJSON) == testMigration012_listingIndexFixture {
+		t.Fatal("listings.json should have different content")
+	}
+
+	var (
+		whitespaceRegex   = regexp.MustCompile(`\s`)
+		hashPropertyRegex = regexp.MustCompile(`"hash":"[a-zA-Z0-9]*",`)
+	)
+
+	listingsIndexJSONStr := whitespaceRegex.ReplaceAllString(string(listingsIndexJSON), "")
+	listingsIndexJSONStr = hashPropertyRegex.ReplaceAllString(listingsIndexJSONStr, "")
+	expectedListingsIndexJSONStr := whitespaceRegex.ReplaceAllString(testMigration012_listingIndexFixture, "")
+	expectedListingsIndexJSONStr = hashPropertyRegex.ReplaceAllString(expectedListingsIndexJSONStr, "")
+
+	if listingsIndexJSONStr != expectedListingsIndexJSONStr {
+		t.Fatal("listings.json should have the same content without hashes")
+	}
+
+	// Check that correct listing's hashes changed
+	listingsIndex := []*migrations.Migration012_ListingData{}
+	err = json.Unmarshal(listingsIndexJSON, &listingsIndex)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expectedHashLengths := []int{0, 0, 0, 49, 49}
```

Comment:
Are these lengths just known quantities are they based on something in the test?

Reply:
It's a known length - the length of a base58 encoded multihash.

Reply:
I think they'll all be 49 but made this check into a looser sanity check that checks the hash length is >= 32.